### PR TITLE
net: Ensure a tmp with exec permissions for dhcp

### PR DESCRIPTION
**Pull request by aciba90**
## Proposed Commit Message

```
net: Ensure a tmp with exec permissions for dhcp

In the case cloudinit.temp_utils points to a fs mounted as noexec
and needs_exe=True, fallback to use
os.join.path(Distro.usr_lib_exec, "cloud-init/clouddir) that
will be mounted with exec perms.

LP: #1962343
```

## Additional Context

https://bugs.launchpad.net/cloud-init/+bug/1962343
https://warthogs.atlassian.net/browse/SC-958

## Test Steps

Execute `tests/integration_tests/datasources/test_tmp_noexec.py`

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/net/ephemeral.py
Comment (holmanb) [inline cloudinit/net/ephemeral.py:443]: Fixing the reversed ipv4/ipv6 context managers should probably be a separate PR. I suspect this technically breaks ephemeral ipv4 on ec2 currently (which is not currently in any releases).

File: cloudinit/distros/__init__.py
Comment (cjp256) [inline cloudinit/distros/__init__.py:939]: Given that this interface is exposed to other packages, consider dropping the leading `_`?

File: cloudinit/net/dhcp.py
Comment (cjp256) [inline cloudinit/net/dhcp.py:45]: Perhaps  needs_exe=True to temp_utils.tmpdir() should just do the right thing and avoid requiring alt_exe_dir passed around?

**Comment by cjp256** [comment]
Thank you for fixing this!! :D

**Review by holmanb — CHANGES_REQUESTED** [review]
Thanks for this @aciba90!

I have one comment about a debug log that I think should be a warning, and a request/suggestion to check that the alt_exe_dir is itself executable. Once that and @cjp256's comments are addressed, I think this will be ready to merge.

File: cloudinit/net/ephemeral.py
Comment (aciba90) [reply cloudinit/net/ephemeral.py:443]: Extracted here: https://github.com/canonical/cloud-init/pull/1697

File: cloudinit/net/dhcp.py
Comment (holmanb) [reply cloudinit/net/dhcp.py:45]: I was wondering about that too (or some other way to reduce the number of repeated changes), however doing that would couple `temp_utils` to `distro`. Is maybe that why @aciba90 didn't do this originally? I don't know off the top of my head if such a change would cause circular import issues, but it might.

File: cloudinit/distros/__init__.py
Comment (aciba90) [reply cloudinit/distros/__init__.py]: Done.

File: cloudinit/net/dhcp.py
Comment (aciba90) [reply cloudinit/net/dhcp.py]: Refactored to expose the target tmp_dir and let callers (datasources, modules) determine if the dir is mounted in a noexec point and give an exec dir via the distro in use.

File: cloudinit/temp_utils.py
Comment (aciba90) [reply cloudinit/temp_utils.py]: Changed to warning. This exposed other uses of `needs_exe=True` without handling the `noexec` condition in:

- cc_chef
- cc_growpart
- cc_puppet
- cc_ubuntu_drivers

**Comment by aciba90** [comment]
Thanks, @cjp256 and @holmanb, for the constructive comments!

I have addressed your comments by exposing a function to let the callers of `cloudinit.temp_utils` know in advance what will be a tmp mount-point, to let them select the correct one.

This is ready to be reviewed after CI passes. Please, let me know if there are more concerns.

**Review by holmanb — CHANGES_REQUESTED** [review]
I really like the integration test addition.

I assume fstab failing to mount on boot would cause the test to fail, but do you think we should add an assertion that ensures the test can't pass if not mounted noexec? If you think this is completely redundant, feel free to push back on this.

File: tests/integration_tests/datasources/test_tmp_noexec.py
Comment (aciba90) [reply tests/integration_tests/datasources/test_tmp_noexec.py:23]: Good idea! Added and retested.

**Comment by aciba90** [comment]
> I really like the integration test addition.
> 
> I assume fstab failing to mount on boot would cause the test to fail, but do you think we should add an assertion that ensures the test can't pass if not mounted noexec? If you think this is completely redundant, feel free to push back on this.

Thanks, @holmanb, for the review and suggestion. I think it is a great idea to assert test preconditions. I have done so and retested it.

File: cloudinit/sources/DataSourceScaleway.py
Comment (cjp256) [inline cloudinit/sources/DataSourceScaleway.py:215]: I think having each caller sort out tmp_dir is going to be error prone.  Even at the lowest level, `tempdir(needs_exe=True) ` now has a hidden dependency on `dir=self.distro.get_tmp_exec_path()`.

I've got two thoughts:
- figure out how to tell temp_utils where the executable tmp dir is without a distro reference on init?
- add a distro.executable_tmpdir() which handles the needs_exe=True and executable dir requirements, dropping those aspects from tmpdir().   this adds a requirement for net to become distro-aware, which I'm actually surprised isn't already the case, but suspect it'll happen eventually as bringing up ephemeral dhcp is likely to have distro-specific aspects (e.g. distros removing dhclient at some point).

Comment (aciba90) [reply cloudinit/sources/DataSourceScaleway.py:215]: Thanks, @cjp256, for the detailed and interesting discussion.

> * figure out how to tell temp_utils where the executable tmp dir is without a distro reference on init?

In the current state of the PR (commit 29833a75611f67d0f86b954bb78f53c17f14eba5), there is no reference to `distro` within `temp_utils`.

> [0]: add a distro.executable_tmpdir() which handles the needs_exe=True and executable dir requirements, dropping those aspects from tmpdir().

There is a new Distro method [get_tmp_exec_path](https://github.com/canonical/cloud-init/pull/1690/files#diff-2902e7f3123dd1e701482d6c60d6ca2e51d2dcbe3f643cb30be70727fde3fbfbR953) which handles the executable dir requirements. It checks if temp_dir with need_exe target tmp dir is in a noexec mount point (via a new function [temp_utils.get_tmp_ancestor](https://github.com/canonical/cloud-init/pull/1690/files#diff-6179e8b68a1cb97cb60b99f1c269a896292fc6dec97cb450cec13886a0101e80R18)) and if so fallback to `{distro.usr_lib_exec}/cloud-init/clouddir`.

We could move the `need_exe` from `temp_utils` to `distro.get_tmp_exec_path` simplifying things a bit.

> [1]: this adds a requirement for net to become distro-aware, which I'm actually surprised isn't already the case, but suspect it'll happen eventually as bringing up ephemeral dhcp is likely to have distro-specific aspects (e.g. distros removing dhclient at some point).

I do not see how [0] implies [1], but assuming it does, I had some arguments against tightly coupling `distro` with `net.dchp` and/or `net.ephemeral`:

1. The necessity of copying and running `dchp` from a tmp dir in order to avoid apparmor restrictions is going to be eliminated (@holmanb can point to links about this, I believe). This implies the dependency will disappear.
2. After [WIP-ONGOING-REFACTORIZATION.rst](https://github.com/canonical/cloud-init/blob/main/WIP-ONGOING-REFACTORIZATION.rst), `net.dchp` and `net.ephemeral` are going to belong to a `cloudinit.distros.networking` hierarchy of classes mapped to the distro hierarchy. If the number of distro-specific aspects (to bring up ephemeral dchp) increases, we could encode the dependencies via this hierarchy.
3. I remember a discussion about dchp. I think `dchp` is going to be discontinued and we talk about the possibility of including a new dchp python-dependency and other alternatives. This will also remove the dependency. (But this belongs to the more obscure future)

Summarizing, I do not think that to only have a distro's exec tmp dir and net justifies to tightly couple them, but I am open to hear more arguments. Maybe @blackboxsw, @holmanb and @TheRealFalcon have opinions/arguments about this.

Comment (holmanb) [reply cloudinit/sources/DataSourceScaleway.py:215]: @aciba90 @cjp256 I'm glad we're thinking about architecture, refactoring, and future direction here. Good points made all around.

The topic of dhcp (both sandboxing issues and the looming issue of dhclient's support status) has come up on irc periodically and maybe also in PR conversations. This topic probably deserves better community coordination and communication, since the problem has implications for clouds and distros (i.e. if we wanted to use a dedicated or auto-assigned link local ipv4 address, dhcp probably wouldn't be necessary, but does that work for all clouds?). This probably deserves a mailing list discussion. Or maybe we should find a way to get contributors in the same room to talk about this kind of stuff :)

> in order to avoid apparmor restrictions is going to be eliminated

This would eliminate some headaches and one historical source of bugs. However we might have to push on this one ourselves.

> Even at the lowest level, tempdir(needs_exe=True) now has a hidden dependency on dir=self.distro.get_tmp_exec_path().

Agreed, we already have implicit coupling.
It seems like more and more code has been shifting into the `distro` implementations recently. I expect this trend will likely continue, which suggests that `distro` may become required by more modules over time.

> 1. The necessity of copying and running dchp from a tmp dir in order to avoid apparmor restrictions is going to be eliminated (@holmanb can point to links about this, I believe). This implies the dependency will disappear.
> ...
> 3. I remember a discussion about dchp. I think dchp is going to be discontinued and we talk about the possibility of including a new dchp python-dependency and other alternatives. This will also remove the dependency. (But this belongs to the more obscure future)

Either of these solutions could potentially remove the requirement for a writable executable tempdir for networking. It does not, however, remove all dependency in cloud-init on `needs_exe`. I see at least a couple of other calls (though the debconf one does look unnecessary at a glance).

All in all, I'm not opposed to this approach. I do think that since we currently need distro input in this patch, it might be worth just creating a distro wrapper function to make the callsites less error prone and more DRY. Whether to do that before this PR lands or as a refactor after, I don't feel strongly about.

That's a lot of words to say basically "I don't have strong feelings on this. I could still be swayed either way." I'll wait in case @cjp256 @blackboxsw @TheRealFalcon or anyone else has more thoughts.

Comment (cjp256) [reply cloudinit/sources/DataSourceScaleway.py:215]: To clarify, I'm not demanding any changes - I'm just driving by and sharing my thoughts fwiw. :D  I'm happy to have this fixed in whatever way you think is best at this time!

Comment (holmanb) [reply cloudinit/sources/DataSourceScaleway.py:215]: @cjp256 Thanks much for sharing. Sounds good :)

File: cloudinit/config/cc_chef.py
Comment (holmanb) [inline cloudinit/config/cc_chef.py:305]: If this branch is taken then on line 315 we will be concatenating a list and a tuple, which will cause a typerror.
```python
_args[0] = [tmpf] + _args[0]
```

**Review by holmanb — CHANGES_REQUESTED** [review]
I just noticed one last nit. It's not something you caused, but since we're modifying this chunk of code I think we should fix it. Pyright picks up error on main (maybe should we supplement mypy), but not on your branch. The problem still exists, however, it just gets hidden from type analyzers in the changes here. I suspect a fix should be as simple as `s/tuple()/list()/g` on line highlighted in the comment.

I'll merge once that issue is fixed (I'd merge now but I want a second opinion on my suggestion).

**Review by holmanb — APPROVED** [review]
Actually it has only one call sight, and we can see that kwargs contains args, so we can simplify a bit. I'm going to merge this and we'll do a little cleanup PR for that shortly.

LGTM. Thanks!

**Comment by cjp256** [comment]
Thanks Alberto @aciba90!

**Comment by aciba90** [comment]
> Thanks Alberto @aciba90!

Thank you, @cjp256, for the reviews and feedback!

**Pull request merged**
